File: packages/core/src/shell/builtins/exit.ts
/**
 * exit - exit the shell
 */

import { BuiltinCommand, ShellEnvironment, ShellOptions } from '../types.js';

export class ExitCommand implements BuiltinCommand {
  name = 'exit';
  description = 'Exit the shell';

  async execute(args: string[], env: ShellEnvironment, options: ShellOptions): Promise<number> {
    let exitCode = env.lastExitCode;

    // If an argument is provided, use it as the exit code
    if (args.length > 0) {
      const code = parseInt(args[0]);
      if (isNaN(code)) {
        console.error(`exit: ${args[0]}: numeric argument required`);
        return 2;
      }
      exitCode = code & 255; // Exit codes are 8-bit
    }

    // Check for stopped jobs in interactive mode
    if (options.interactiveMode && env.jobs) {
      const stoppedJobs = Array.from(env.jobs.values())
        .filter(job => job.status === 'stopped');
      
      if (stoppedJobs.length > 0) {
        console.error('exit: There are stopped jobs.');
        // In a real shell, this would prevent exit on first attempt
        // For now, we'll just warn
      }
    }

    // Exit the process
    process.exit(exitCode);
  }

  help(): string {
    return `exit [n]
Exit the shell with a status of N.
If N is omitted, the exit status is that of the last command executed.

The exit status is limited to the range 0-255.`;
  }
}

File: shell/src/shell/builtins/exit.rs
//! exit - exit the shell builtin command

use crate::shell::types::*;

pub struct ExitCommand;

#[async_trait::async_trait]
impl BuiltinCommand for ExitCommand {
    fn name(&self) -> &str {
        "exit"
    }

    fn description(&self) -> &str {
        "Exit the shell"
    }

    async fn execute(
        &self,
        args: &[String],
        env: &mut ShellEnvironment,
        _options: &ShellOptions,
    ) -> Result<i32, ShellError> {
        let exit_code = if args.is_empty() {
            // Use last exit code if no argument given
            env.last_exit_code
        } else {
            // Parse the provided exit code
            match args[0].parse::<i32>() {
                Ok(code) => code,
                Err(_) => {
                    eprintln!("exit: {}: numeric argument required", args[0]);
                    return Ok(128); // Exit with error
                }
            }
        };

        // For now, we'll use a special error to indicate shell should exit
        Err(ShellError::RuntimeError {
            message: format!("EXIT:{}", exit_code),
        })
    }

    fn help(&self) -> String {
        "exit [n]
Exit the shell with a status of N.
If N is omitted, the exit status is that of the last command executed.".to_string()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_exit_no_args() {
        let mut env = ShellEnvironment::new();
        env.last_exit_code = 42;
        
        let exit = ExitCommand;
        let result = exit.execute(&[], &mut env, &ShellOptions::default()).await;
        
        assert!(result.is_err());
        match result.unwrap_err() {
            ShellError::RuntimeError { message } => {
                assert_eq!(message, "EXIT:42");
            }
            _ => panic!("Expected RuntimeError with EXIT message"),
        }
    }

    #[tokio::test]
    async fn test_exit_with_code() {
        let mut env = ShellEnvironment::new();
        let exit = ExitCommand;
        
        let result = exit.execute(
            &["5".to_string()], 
            &mut env, 
            &ShellOptions::default()
        ).await;
        
        assert!(result.is_err());
        match result.unwrap_err() {
            ShellError::RuntimeError { message } => {
                assert_eq!(message, "EXIT:5");
            }
            _ => panic!("Expected RuntimeError with EXIT message"),
        }
    }

    #[tokio::test]
    async fn test_exit_invalid_code() {
        let mut env = ShellEnvironment::new();
        let exit = ExitCommand;
        
        let result = exit.execute(
            &["invalid".to_string()], 
            &mut env, 
            &ShellOptions::default()
        ).await;
        
        assert!(result.is_ok());
        assert_eq!(result.unwrap(), 128);
    }
}